src/anyvar/storage/duckdb.py

```diff
@@ -0,0 +1,226 @@
+        :param value: value for `vrs_object` field
+        """
+        # Convert the value to a JSON string
+        value_json = json.dumps(value.model_dump(exclude_none=True))
+
+        # Use INSERT with ON CONFLICT DO NOTHING
+        insert_query = f"""
+            INSERT INTO {self.table_name} (vrs_id, vrs_object)
+            VALUES (?, ?)
+            ON CONFLICT DO NOTHING
+        """  # noqa: S608
+
+        # Execute the query with parameterized values
+        db_conn.execute(insert_query, (name, value_json))
+
+    def _random_tmp_table_name(self) -> str:
+        return "".join(random.choice(string.ascii_uppercase) for i in range(32))
+
+    def add_many_items(
+        self,
+        db_conn: duckdb.DuckDBPyConnection,
+        items: list[tuple[str, pydantic.BaseModel]],
+    ) -> None:
+        """Perform bulk insert using a temporary table in DuckDB.
+
+        :param db_conn: a DuckDB database connection
+        :param items: list of tuples (name, value) to be inserted
+        """
+        # Create a temporary table with the same schema as the main table
+        # TODO if application has any concurrency, tmp_table name should be made unique instead
+        # Create random name starting with tmp_table
+        tmp_table_name = f"tmp_table_{self._random_tmp_table_name()}"
+        tmp_statement = f"""
+            CREATE TEMPORARY TABLE {tmp_table_name} (vrs_id TEXT, vrs_object JSON)
+        """
+        insert_statement = f"""
+            INSERT INTO {self.table_name}
+            SELECT * FROM {tmp_table_name}
+            ON CONFLICT (vrs_id) DO NOTHING
+        """  # noqa: S608
+        drop_statement = f"DROP TABLE {tmp_table_name}"
+
+        # Create the temporary table
+        db_conn.execute(tmp_statement)
+
+        # Prepare data for bulk insertion using a Pandas DataFrame
+        row_data = [
+            (name, json.dumps(value.model_dump(exclude_none=True)))
+            for name, value in items
+        ]
+
+        # Insert data into the temporary table
+        db_conn.execute(
+            f"INSERT INTO {tmp_table_name} VALUES (?, ?)", row_data  # noqa: S608
+        )
+
+        # Move data from the temporary table to the main table with conflict handling
+        db_conn.execute(insert_statement)
+
+        # Drop the temporary table
+        db_conn.execute(drop_statement)
+
+    def deletion_count(self, db_conn: Connection) -> int:
+        """Count the number of VRS objects with no sequence
+
+        :param db_conn: a database connection
+        :param vrs_id: the VRS ID
+        """
+        result = db_conn.execute(
+            sql_text(
+                f"""
+            SELECT COUNT(*) AS c
+              FROM {self.table_name}
+             WHERE LENGTH(vrs_object -> 'state' ->> 'sequence') = 0
+            """  # noqa: S608
+            )
+        )
+        return result.scalar()
+
+    def substitution_count(self, db_conn: Connection) -> int:
```

Comment:
just a side rumination. do we really need all these count methods as part of the interface?

Reply:
No, and https://github.com/biocommons/anyvar/pull/118 removes them from the `SqlStorage` class